jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/ProjectProperties.java

```diff
@@ -30,45 +33,91 @@
 /** Obtains information about a Gradle {@link Project} that uses Jib. */
 class ProjectProperties {
 
+  private static final String PLUGIN_NAME = "jib";
+
   private final Project project;
   private final GradleBuildLogger gradleBuildLogger;
+  private final SourceFilesConfiguration sourceFilesConfiguration;
+
+  /** @return a ProjectProperties from the given project and logger. */
+  static ProjectProperties getForProject(Project project, GradleBuildLogger gradleBuildLogger) {
+    try {
+      return new ProjectProperties(
+          project, gradleBuildLogger, GradleSourceFilesConfiguration.getForProject(project));
+    } catch (IOException ex) {
+      throw new GradleException("Obtaining project build output files failed", ex);
+    }
+  }
 
-  ProjectProperties(Project project, GradleBuildLogger gradleBuildLogger) {
+  @VisibleForTesting
+  ProjectProperties(
+      Project project,
+      GradleBuildLogger gradleBuildLogger,
+      SourceFilesConfiguration sourceFilesConfiguration) {
     this.project = project;
     this.gradleBuildLogger = gradleBuildLogger;
+    this.sourceFilesConfiguration = sourceFilesConfiguration;
   }
 
   /**
-   * @param mainClass the configured main class
-   * @return the main class to use for the container entrypoint.
+   * If {@code mainClass} is {@code null}, tries to infer main class in this order:
+   *
+   * <ul>
+   *   <li>1. Looks in a {@code jar} task.
+   *   <li>2. Searches for a class defined with a main method.
+   * </ul>
+   *
+   * <p>Warns if main class is not valid.
+   *
+   * @throws GradleException if no valid main class is not found.
    */
   String getMainClass(@Nullable String mainClass) {
     if (mainClass == null) {
+      gradleBuildLogger.info(
+          "Searching for main class... Add a 'mainClass' configuration to '"
+              + PLUGIN_NAME
+              + "' to improve build speed.");
       mainClass = getMainClassFromJarTask();
       if (mainClass == null) {
-        throw new GradleException(
-            HelpfulSuggestionsProvider.get("Could not find main class specified in a 'jar' task")
-                .suggest("add a `mainClass` configuration to jib"));
+        gradleBuildLogger.debug(
+            "Could not find main class specified in a 'jar' task; attempting to "
```

Comment:
Let's add a todo to consolidate this code with the Maven one.

Reply:
Already added it to #262, but I can add a TODO in the code as well if you want.

Reply:
Okay that's fine then.